CommonTools/Utils/test/testFormulaEvaluator.cc

```diff
@@ -585,6 +585,214 @@ void testFormulaEvaluator::checkFormulaEvaluator() {
     CPPUNIT_ASSERT(f.evaluate(emptyV, emptyV) == 2);
   }
 
+  {
+    reco::FormulaEvaluator f("cos(0.5)");
+
+    std::vector<double> emptyV;
+
+    CPPUNIT_ASSERT(fabs(f.evaluate(emptyV, emptyV) - std::cos(0.5)) < 1e-9);
```

Comment:
similarly complex log, pow, exp unit tests are using equality.
Please switch to it here (for all of the added tests) unless there is a strong reason not to.

Reply:
I believe the trigger for this is that the tests for some of the functions failed right above commit https://github.com/cms-sw/cmssw/pull/27987/commits/094b3f6f09194c982daaab0de22e72bfee35db76 .

It is traced down to a difference at the 16th digit after the decimal point for std::acosh(...).  What would you suggest them to do?

Best Regards,
Yi Chen

Reply:
All other functions have passed the equality check in the previous test except for std::acosh(...), and the cause of the failure is as mentioned by Yi Chen. Upon further inspection, the std::atanh(...) have the same issue too.

An example of the issue is as demonstrated below:

    std::cout << std::setprecision(20) << "acosh(2.0) = " << f.evaluate(emptyV, emptyV) << "\n";
    std::cout << std::setprecision(20) << "std::acosh(2.0) = " << std::acosh(2.0) << "\n";
    std::cout << "acosh(2.0) - std::acosh(2.0) = " << fabs(f.evaluate(emptyV, emptyV) - std::acosh(2.0));

The results:

> .....acosh(2.0) = 1.3169578969248165734
> std::acosh(2.0) = 1.3169578969248167954
> acosh(2.0) - std::acosh(2.0) = 2.2204460492503130808e-16

Hence I change the requirement for the test from evaluating equality to the difference (fabs(a-b) < 1e-9...), and for consistency, I did the same for every other trigonometric function added.

What is your suggestion to resolve this? Should I maintain using equality for every other function and only calculate the difference for std::acosh(...) and std::atanh(...)?

Reply:
Note that this only happens for std::acosh() and std::atanh() but no problem for the corresponding TMath functions.

Reply:
for all cases where equality works, I prefer that we use equality.
It would be good to understand the origins of the difference in the other case, but for the unit test now an abs seems to be the only solution.

Reply:
@Dr15Jones 
please check if you have some insight in the numerical difference observed for achosh calls.
Thank you.

Reply:
gcc appears to be calculating the value at compile time, see https://godbolt.org/z/4sADhe, and gets a different value from the call.  While if I do the same with clang, see https://godbolt.org/z/4KmgvL, it calls std::acosh and returns the same value as the call to the modified FormulaEvaluator. 

So this is just a case of two different ways of truncating a double at the last bit.

Reply:
should we try to convince gcc here to compute the reference value at run time
or is is reasonable to just leave fabs(delta) here?

Reply:
How about leave the delta but add a comment about gcc computing the value differently at compile time?

Reply:
Thanks @Dr15Jones for the insight. So how do I proceed, @slava77? Comment them out?

Reply:
> Thanks @Dr15Jones for the insight. So how do I proceed, @slava77? Add comment above the deltas?

yes, please add a comment above the ```CPPUNIT_ASSERT``` lines using fabs [one comment is enough if these fabs are within 5-10 lines from each other]